fedimint-core/src/util/mod.rs

```diff
@@ -465,18 +466,56 @@ where
                 } else {
                     warn!(
                         target: LOG_CORE,
-                        ?error,
+                        ?err,
                         %attempts,
                         "{} failed",
                         op_name,
                     );
-                    return Err(error);
+                    return Err(err);
                 }
             }
         }
     }
 }
 
+/// A wrapper with `fmt::Display` for any `E : Error` that will print chain of
+/// causes
+pub struct WithCause<'e, E>(pub &'e E);
+
+impl<E> fmt::Display for WithCause<'_, E>
+where
+    E: error::Error,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut source_opt: Option<_> = Some(&self.0);
+
+        while source_opt.is_some() {
+            let source = source_opt.take().expect("Just checked");
+            f.write_fmt(format_args!("{source}"))?;
+
+            let source = source.source();
+            if source.is_some() {
+                f.write_str(": ")?;
+            }
+        }
+        Ok(())
```

Comment:
this while doesn't loop? it only works as `if`

Reply:
@maan2003 Am I missing something? I should add a test...

Oh damn, it was supposed to be `source = source.source();`... 

Good catch :D